File: LUA/Cap08/c08ex02.lua
-- c08ex02.lua

tn = tonumber

function caractere_especial(entra)
  local caractere_especial = "!@#$%^&*()-_=+[]{}|;:',<>/?`~\""
  for i = 1, #caractere_especial do
    if string.find(entra, "%"..caractere_especial:sub(i,i)) then
      return true
    end
  end
  return false
end

function valida_numero(entra)
  local valor = tn(entra)
  return valor ~= nil
end

while true do
  io.write("Entre um valor numerico: ")
  local numero = io.read()
  print()

  if (numero == "") then
    print("Erro: Tecla <Enter> acionada acidentalmente.\n")
  else
    if (caractere_especial(numero)) then
      print("Erro: Caractere especial nao permitido.\n")
    elseif (valida_numero(numero)) then
      print("Entrada valida fornecida: " .. numero .. "\n")
      break
    else
      print("Erro: Entre um numero valido (inteiro ou ponto flutuante).\n")
    end
  end
end

io.write("\nTecle <Enter> para encerrar... ")
io.read()

File: PYTHON/Cap08/c08ex02.py
# c08ex02.py

def caractere_especial(entra):
  caractere_especial = "!@#$%^&*()-_=+[]{}|;:',<>/?`~\""
  for i in caractere_especial:
    if i in entra:
      return True
  return False

def valida_numero(entra):
  try:
    float(entra)
    return True
  except ValueError:
    return False

while True:
  print("Entre um valor numerico: ", end="")
  numero = input()
  print()

  if not numero:
    print("Erro: Tecla <Enter> acionada acidentalmente.\n")
  elif caractere_especial(numero):
    print("Erro: Caractere especial nao permitido.\n")
  elif valida_numero(numero):
    print(f"Entrada valida fornecida: {numero}\n")
    break
  else:
    print("Erro: Entre um numero valido (inteiro ou real).\n")

input("\nTecle <Enter> para encerrar... ")
